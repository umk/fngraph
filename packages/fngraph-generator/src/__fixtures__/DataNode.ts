import DataNode from '../DataNode'

export function createDataNode(overrides?: Partial<DataNode>): DataNode {
  return {
    incoming: [],
    outgoing: [],
    getter: jest.fn(),
    priority: 0,
    invert: false,
    getProperties: () => [],
    ...overrides,
  }
}
