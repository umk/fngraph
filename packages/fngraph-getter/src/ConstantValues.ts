import { DataRecord } from '@fngraph/data'
import {
  CONTEXT_GROUP_INVERSION,
  Falsy,
  GeneratorValue,
  Getter,
  GetterFactory,
} from '@fngraph/generator'

import { createIteration } from './IterationFunction'

type ConstantValues<V extends DataRecord> = Array<V | Falsy> | V | Falsy

export function createConstant<V extends DataRecord>(
  source: ConstantValues<V>,
): GetterFactory<never, V> {
  return function (_incoming, outgoing): Getter {
    const iterationF = createIteration(outgoing)
    return async function* (
      _properties,
      records,
      context,
      getRecord,
      invert,
    ): AsyncGenerator<GeneratorValue> {
      const iteration = iterationF(context, invert)
      for await (const current of records) {
        context.advanceToValueGroup(current)
        const base = getRecord(current.contexts)
        if (base) {
          const valuesArr = Array.isArray(source) ? source : [source]
          if (!invert) {
            yield* iteration(current, base, valuesArr.filter(Boolean) as Array<V>)
          } else {
            if (!valuesArr.some(Boolean)) {
              yield* iteration(current, base, CONTEXT_GROUP_INVERSION as V)
            }
          }
        }
      }
    }
  }
}

export default ConstantValues
