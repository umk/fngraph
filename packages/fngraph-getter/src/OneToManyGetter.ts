import { DataRecord } from '@fngraph/data'
import {
  CONTEXT_GROUP_INVERSION,
  ContextArrayGroup,
  Falsy,
  GeneratorValue,
  Getter,
  GetterFactory,
  PropertyRef,
} from '@fngraph/generator'

import { createIteration } from './IterationFunction'

type OneToManyGetter<P extends DataRecord, R extends DataRecord> = (
  record: P,
  properties: Array<PropertyRef>,
) => AsyncGenerator<R | Falsy>

export function createOneToMany<P extends DataRecord, R extends DataRecord>(
  source: OneToManyGetter<P, R>,
): GetterFactory<P, R> {
  return function (incoming, outgoing): Getter {
    const iterationF = createIteration(outgoing)
    return async function* (
      properties,
      records,
      context,
      getRecord,
      invert,
    ): AsyncGenerator<GeneratorValue> {
      const iteration = iterationF(context, invert)
      async function* getSourceRecords(current: GeneratorValue) {
        if (context.groups.length > context.groupIndex) {
          const group = context.groups[context.groupIndex] as ContextArrayGroup<R> | undefined
          if (group) {
            // The cached derived record is defined only if and only
            // if the context is also defined.
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            const base = getRecord(current.contexts)!
            yield { base, group }
          }
        } else {
          const base = getRecord(current.contexts)
          if (base) {
            const derivations = source(incoming(base), properties)
            let group: ContextArrayGroup<R> = []
            if (context.isCached) {
              group = []
              context.groups.push(group)
            }
            if (!invert) {
              for await (const derived of derivations) {
                if (derived) group.push(derived)
              }
            } else {
              for await (const derived of derivations) {
                if (derived) return
              }
              group.push(CONTEXT_GROUP_INVERSION)
            }
            yield { base, group }
          } else {
            context.groups.push(undefined)
          }
        }
        return undefined
      }
      for await (const current of records) {
        context.advanceToValueGroup(current)
        const records = getSourceRecords(current)
        for await (const record of records) {
          yield* iteration(current, record.base, record.group as Array<R>)
        }
        context.groupIndex++
      }
    }
  }
}

export default OneToManyGetter
