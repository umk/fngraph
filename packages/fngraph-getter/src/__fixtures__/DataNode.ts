import { DataNode } from '@fngraph/generator'

export function createDataNode(overrides?: Partial<DataNode>): DataNode {
  return {
    incoming: [],
    outgoing: [],
    getter: jest.fn(),
    priority: 0,
    invert: false,
    getProperties: () => [],
    ...overrides,
  }
}
