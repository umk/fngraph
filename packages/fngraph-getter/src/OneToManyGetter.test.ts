import { DataRecord, DeclarationID } from '@fngraph/data'
import { contextAsRecord, createDataRecordGenerator, recordAsContext } from '@fngraph/generator'

import { createConstant } from './ConstantValues'
import { createOneToMany } from './OneToManyGetter'
import { createDataNode } from './__fixtures__'

describe('createOneToMany', () => {
  const declarationA = 'declaration A' as DeclarationID
  const declarationB = 'declaration B' as DeclarationID
  const declarationC = 'declaration C' as DeclarationID
  const declarationD = 'declaration D' as DeclarationID
  it('iterates through getter records', async () => {
    const nodeA = createDataNode({
      outgoing: [declarationA, declarationB],
      getter: createConstant([
        { [declarationA]: 1, [declarationB]: 2 },
        { [declarationA]: 10, [declarationB]: 20 },
      ])(contextAsRecord, recordAsContext),
    })
    const nodeB = createDataNode({
      incoming: [declarationA, declarationB],
      outgoing: [declarationC],
      getter: createOneToMany((record) => {
        const valueA = record[declarationA] as number
        const valueB = record[declarationB] as number
        return (async function* () {
          yield { [declarationC]: valueA + valueB }
          yield { [declarationC]: valueA - valueB }
        })()
      })(contextAsRecord, recordAsContext),
    })
    const generator = createDataRecordGenerator([nodeA, nodeB], [])
    const records: Array<DataRecord> = []
    for await (const record of generator()) records.push(record)
    expect(records).toMatchSnapshot()
  })
  it('permutates records with another node', async () => {
    /**
     * Must result in the following evaluation sequence:
     *
     *   (B) <- (C) <- (A)
     *
     *        - or -
     *
     *   (B) <- (A) <- (C)
     */
    const nodeA = createDataNode({
      outgoing: [declarationA, declarationB],
      getter: createConstant([
        { [declarationA]: 1, [declarationB]: 2 },
        { [declarationA]: 3, [declarationB]: 4 },
      ])(contextAsRecord, recordAsContext),
    })
    const getter = jest.fn((record) => {
      const valueA = record[declarationA] as number
      const valueB = record[declarationB] as number
      return (async function* () {
        yield { [declarationC]: valueA + valueB }
        yield { [declarationC]: valueA - valueB }
      })()
    })
    const nodeB = createDataNode({
      incoming: [declarationA, declarationB],
      outgoing: [declarationC],
      getter: createOneToMany(getter)(contextAsRecord, recordAsContext),
    })
    const nodeC = createDataNode({
      outgoing: [declarationD],
      getter: createConstant([{ [declarationD]: 10 }, { [declarationD]: 20 }])(
        contextAsRecord,
        recordAsContext,
      ),
    })
    const generator = createDataRecordGenerator([nodeA, nodeB, nodeC], [])
    const records: Array<DataRecord> = []
    for await (const record of generator()) records.push(record)
    expect(records).toMatchSnapshot()
    expect(getter).toHaveBeenCalledTimes(2)
  })
})
