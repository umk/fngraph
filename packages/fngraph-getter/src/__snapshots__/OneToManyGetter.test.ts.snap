// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createOneToMany iterates through getter records 1`] = `
[
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": 3,
  },
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": -1,
  },
  {
    "declaration A": 10,
    "declaration B": 20,
    "declaration C": 30,
  },
  {
    "declaration A": 10,
    "declaration B": 20,
    "declaration C": -10,
  },
]
`;

exports[`createOneToMany permutates records with another node 1`] = `
[
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": 3,
    "declaration D": 10,
  },
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": -1,
    "declaration D": 10,
  },
  {
    "declaration A": 3,
    "declaration B": 4,
    "declaration C": 7,
    "declaration D": 10,
  },
  {
    "declaration A": 3,
    "declaration B": 4,
    "declaration C": -1,
    "declaration D": 10,
  },
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": 3,
    "declaration D": 20,
  },
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": -1,
    "declaration D": 20,
  },
  {
    "declaration A": 3,
    "declaration B": 4,
    "declaration C": 7,
    "declaration D": 20,
  },
  {
    "declaration A": 3,
    "declaration B": 4,
    "declaration C": -1,
    "declaration D": 20,
  },
]
`;
